tests/ptycho/test_ptycho.py

```diff
@@ -398,6 +482,86 @@ def test_consistent_lstsq_grad_variable_probe(self):
             "These weights should be unused/untouched "
             "and should have been initialized to zero.")
+        
+    #===========
+    # rPIE tests
+    #===========
+
+    def test_consistent_rpie_poisson_steplength_dominantmode(self):
+        """Check ptycho.solver.rpie for consistency."""
+        params = tike.ptycho.PtychoParameters(
+            psi=self.psi,
+            probe=self.probe,
+            scan=self.scan,
+            algorithm_options=tike.ptycho.RpieOptions(
+                num_batch=5,
+                num_iter=16,
+            ),
+            probe_options=ProbeOptions(),
+            object_options=ObjectOptions(),
+            exitwave_options=ExitWaveOptions( noise_model = "poisson", step_length_usemodes = "dominant_mode" ),
+        )
+        _save_ptycho_result(
+            self.template_consistent_algorithm(
+                data=self.data,
+                params=params,
+            ),
+            f"mpi{self.mpi_size}-rpie_poisson_steplength_dominantmode{self.post_name}",
+        )
+
+    def test_consistent_rpie_poisson_steplength_allmodes(self):
+        """Check ptycho.solver.rpie for consistency."""
+        params = tike.ptycho.PtychoParameters(
+            psi=self.psi,
+            probe=self.probe,
+            scan=self.scan,
+            algorithm_options=tike.ptycho.RpieOptions(
+                num_batch=5,
+                num_iter=16,
+            ),
+            probe_options=ProbeOptions(),
+            object_options=ObjectOptions(),
+            exitwave_options=ExitWaveOptions( noise_model = "poisson", step_length_usemodes = "all_modes" ),
+        )
+        _save_ptycho_result(
+            self.template_consistent_algorithm(
+                data=self.data,
+                params=params,
+            ),
+            f"mpi{self.mpi_size}-rpie_poisson_steplength_allmodes{self.post_name}",
+        )
+
+    def test_consistent_rpie_unmeasured_detector_regions(self):
+        """Check ptycho.solver.rpie for consistency."""
+        params = tike.ptycho.PtychoParameters(
+            psi=self.psi,
+            probe=self.probe,
+            scan=self.scan,
+            algorithm_options=tike.ptycho.RpieOptions(
+                num_batch=5,
+                num_iter=16,
+            ),
+            probe_options=ProbeOptions(),
+            object_options=ObjectOptions(),
+            exitwave_options=ExitWaveOptions( unmeasured_pixels_scaling = 0.90 ),
+        )
+
+        # Define regions where we have missing diffraction measurement data
+        unmeasured_pixels = np.zeros( self.probe.shape[-2:], np.bool_ )
+        unmeasured_pixels[ 100 : 105, : ] = True
+        unmeasured_pixels[ :, 100 : 105 ] = True
+        params.exitwave_options.measured_pixels = np.logical_not( unmeasured_pixels )
+
+        # Zero out these regions on the diffraction measurement data
+        self.data = self.data * params.exitwave_options.measured_pixels
```

Comment:
This shouldn't be a required pre-processing step in order for the detector masking feature to work correctly. We want the user to need to do the least-amount of work to use a feature. We should be able to implement this feature to work correctly with only one mask indicating where the measured pixels are located.

Reply:
I'm just doing that on line 555/556 to *simulate* unmeasured pixels in the measurements.

Upon loading the data from the h5 file, the "data" variable *should* already have he unmeasured regions zeroed out.

Reply:
> Upon loading the data from the h5 file, the "data" variable should already have the unmeasured regions zeroed out.

Experimentalists don't use zero to indicate unmeasured regions because zero is a valid number of photon counts. My point is that adding this extra preprocessing step is redundant because we have a mask of "good pixels" and it is ambiguous because zero is a valid photon count. If you want to test a filler value, you should be using NaN because the NaNs will propagate if you use them to compute anything.

```suggestion
        # Zero out these regions on the diffraction measurement data
        self.data[params.exitwave_options.measured_pixels] = np.nan
```